delta-app/src/main/java/io/cdap/delta/store/RemoteHttpClient.java

```diff
@@ -0,0 +1,154 @@
+  }
+
+  public String retryableApiCall(String connectionUrl, String ...method) {
+    long delay = RETRY_BASE_DELAY_MILLIS;
+    double minMultiplier = RETRY_DELAY_MULTIPLIER - RETRY_DELAY_MULTIPLIER * RETRY_RANDOMIZE_FACTOR;
+    double maxMultiplier = RETRY_DELAY_MULTIPLIER + RETRY_DELAY_MULTIPLIER * RETRY_RANDOMIZE_FACTOR;
+    Stopwatch stopWatch = new Stopwatch().start();
+
+    try {
+      while (stopWatch.elapsedTime(TimeUnit.MILLISECONDS) < TIMEOUT_MILLIS) {
+        try {
+          if (method[0].equals(HttpMethod.GET)) {
+            return executeGetCall(getHttpUrlConnection(connectionUrl));
+          } else if (method[0].equals(HttpMethod.POST) || method[0].equals(HttpMethod.PUT)) {
+            return executeCallWithBody(getHttpUrlConnection(connectionUrl), method[0], method[1]);
+          }
+        } catch (RetryableException | IOException e) {
+          TimeUnit.MILLISECONDS.sleep(delay);
+          delay = (long) (delay * (minMultiplier + Math.random() * (maxMultiplier - minMultiplier + 1)));
+          delay = Math.min(delay, RETRY_MAX_DELAY_MILLIS);
+        }
+      }
+    } catch (InterruptedException e) {
+      throw new RuntimeException("Thread interrupted while trying to connect with State Store Service " +
+                                   "(AssessorService)", e);
+    }
+    throw new IllegalStateException("Timed out when trying to connect with State Store Service (AssessorService)");
+  }
+
+  private String executeGetCall(HttpURLConnection urlConn) throws IOException {
+    urlConn.setRequestMethod(HttpMethod.GET);
+    return validateAndRetrieveContent(urlConn);
+  }
+
+  private String executeCallWithBody(HttpURLConnection urlConn, String method, String body) throws IOException {
+
+    urlConn.setDoOutput(true);
+    urlConn.setRequestMethod(method);
+
+    OutputStream os = urlConn.getOutputStream();
+    OutputStreamWriter osw = new OutputStreamWriter(os, "UTF-8");
+    osw.write(body);
+    osw.flush();
+    osw.close();
+    os.close();
+    urlConn.connect();
+
+    return validateAndRetrieveContent(urlConn);
+  }
+
+  /**
+   * Copied from "AbstractServiceRetryableMacroEvaluator"
+   * Returns the full content of the error stream for the given {@link HttpURLConnection}.
+   */
+  private String validateAndRetrieveContent(HttpURLConnection urlConn) throws IOException {
+    String serviceName = "State Store (AssessorService)";
+    if (urlConn == null) {
+      throw new RetryableException(serviceName + " service is not available");
+    }
+
+    int responseCode = urlConn.getResponseCode();
+    if (responseCode != HttpURLConnection.HTTP_OK) {
+      throw new RetryableException("Failed to call " + serviceName + " service with status " + responseCode + ": " +
+                                     getError(urlConn));
+    }
+
+    try (Reader reader = new InputStreamReader(urlConn.getInputStream(), StandardCharsets.UTF_8)) {
+      return CharStreams.toString(reader);
```

Comment:
returning a String and then parsing that string is not as efficient as decoding it into an object directly from the stream. See the Gson methods that take a Reader as input.

It would be better if the return type of this method is a generic, with the Type passed in as an argument.

Reply:
only `OffsetAndSequence` gson conversion is done within replication App, and it happens once on every start of worker. 

Also tried the above approach the type `byte[].class` for other states is giving issues.